pymeasure/instruments/newport/smc100.py

```diff
@@ -0,0 +1,709 @@
+        "3C": "DISABLE from READY",
+        "3D": "DISABLE from MOVING",
+        "3E": "DISABLE from JOGGING",
+        "46": "JOGGING from READY",
+        "47": "JOGGING from DISABLE"
+    }
+
+    def __init__(self, code):
+        self.code = code
+        self.message = self.MESSAGES[self.code]
+        self.ready = self.message.startswith("READY")
+        self.disable = self.message.startswith("DISABLE")
+        self.not_referenced = self.message.startswith("NOT REFERENCED")
+
+    def __str__(self):
+        return self.message
+
+
+class SMC100(Instrument):
+    """Represents the Newport SMC100 Motion Controller and provides a
+    high-level interface for interacting with the instrument.
+
+    Each SMC100 controls exactly one stage. The first controller is
+    connected to the computer via RS232C, optionally adapted to USB.
+    Other SMC100s can be chained to the first one via the internal RS485
+    link. In this case, each one must have its own address. Refer to
+    User Manual 3.3 for details.
+
+    Refer to the User Manual for details on the functions. One should
+    pay attention to the state of the controller when setting a
+    parameter. This is especially true in the CONFIGURATION state.
+
+    In the description of the properties, u stands for the preset unit.
+    Default is mm and degrees for Newport stages.
+    
+    Three commands can be broadcast to all chained controllers:
+    - Enter/Leave DISABLE state
+    - Execute simultaneous started move
+    - Stop motion
+    These can be implemented by creating an SMC100 instance with
+    address=0. Be carefull: error checking is disabled for broadcast
+    commands. Broadcasting wrong commands will show delayed errors when
+    executing next addressed command.
+
+    Stepper motor, analog and TTL inputs, and keypad related functions
+    are not implemented as they could not be tested. Setting address
+    command is not implemented as it is discouraged in the User Manual.
+    Simultaneous started move are not implemented because it's getting
+    late.
+
+    This code should not be trusted too much, and extensive tests must
+    be ran before executing critical moves.
+
+    Example use:
+
+    .. code-block:: python
+
+        from pymeasure.instruments.newport import SMC100
+
+        with SMC100("ASRL7::INSTR") as smc:
+            smc.search_home()
+            smc.wait_ready()
+            smc.position = 10
+            smc.wait_ready()
+            smc.position += 5.5
+            smc.stop()
+        print(smc.state)
+        print(smc)
+        smc.reset()
+    """
+
+    #Class properties
+    acceleration = Instrument.control(
+        "AC?", "AC%g",
+        """Get or set the acceleration.
+        
+        Type: Floating point number
+        Range: ]1e-6, 1e12[
+        Unit: u/s2""",
+        validator=strict_range,
+        values=(1e-6, 1e12),
+        check_get_errors=True,
+        check_set_errors=True
+    )
```

Comment:
I think that for this instrument, it is interesting to keep systematic error checking by default. If I'm not mistaken, there is no handshake in RS232. Checking command errors grantees that the motion controller got the command, with the added benefit of getting a diagnosis in case of wrongful command.

When I wrote the code, I thought of adding a parameter `check_errors=True` to the `SMC100` class, but it couldn't be used by the class attributes anyway. An alternative would be to add such an option to the `Instrument` class, but I don't know if it would be useful to enough instruments.

Reply:
I understand your point regarding catching errors synchronously, the manual says the following:
"For a safe program flow it is recommended to always query the command error after 
each command execution." If this is the case, you could extend the write method as follow:
```
    def write(self, command):
        super().write(self.address + command)
        # Check error with something similar to this
        super().write(self.address + "TE")
        .
        .

```

Also this should be done similarly for `ask` and `values` methods.
The benefit would be that all error checking will be concentrated in single point and you do not need to repeat the `check_g(s)et_errors=True` or every coommand.
Another idea, could be to locally define some functions like:
```
def checked_control(*args, **kwargs):
     kwargs.setdefault('check_set_errors', True)
     kwargs.setdefault('check_get_errors', True)
     return Instrument.control(*args, **kwargs)

def checked_setting(*args, **kwargs):
     kwargs.setdefault('check_set_errors', True)
     kwargs.setdefault('check_get_errors', True)
     return Instrument.setting(*args, **kwargs)

def checked_measurement(*args, **kwargs):
     kwargs.setdefault('check_set_errors', True)
     kwargs.setdefault('check_get_errors', True)
     return Instrument.measurement(*args, **kwargs)
```
and use them when an automatically checked command need to be defined.

I am interested in your opinion and also other maintainers opinion regarding this approach.

Reply:
I understand your point regarding catching errors synchronously, the manual says the following:
"For a safe program flow it is recommended to always query the command error after 
each command execution." If this is the case, you could extend the write method as follow:
```
    def write(self, command):
        super().write(self.address + command)
        # Check error with something similar to this
        super().write(self.address + "TE")
        .
        .

```

Also this should be done similarly for `ask` and `values` methods.
The benefit would be that all error checking will be concentrated in single point and you do not need to repeat the `check_g(s)et_errors=True` or every coommand.
Another idea, could be to locally define some functions like:
```
def checked_control(*args, **kwargs):
     kwargs.setdefault('check_set_errors', True)
     kwargs.setdefault('check_get_errors', True)
     return Instrument.control(*args, **kwargs)

def checked_setting(*args, **kwargs):
     kwargs.setdefault('check_set_errors', True)
     kwargs.setdefault('check_get_errors', True)
     return Instrument.setting(*args, **kwargs)

def checked_measurement(*args, **kwargs):
     kwargs.setdefault('check_set_errors', True)
     kwargs.setdefault('check_get_errors', True)
     return Instrument.measurement(*args, **kwargs)
```
and use them when an automatically checked command need to be defined.

I am interested in your opinion and also other maintainers opinion regarding this approach.

Reply:
I think repetition of a factory argument is fine, it keeps all relevant information on a property local (in the property definition). 
The `checked_*` factories _are_ a nice solution, though! I would not add them to the base `Instrument`, though, but keep them local. Not sure if that's possible without introducing a base class one layer above SMC100.

Reply:
@bilderbuchi, yes definitely locally.
Being the `Instrument` methods staticmethods (i.e. simply functions), my idea/proposal was to define some functions locally without being part of a specific class.

Reply:
In my previous message, I wrote:

> An alternative would be to add such an option [a `check_error_default` at the class level] to the Instrument class, but I don't know if it would be useful to enough instruments.

I had not really thought it through, though, and it seams that the use of static methods and class attributes make this impossible.

As @msmttchr  suggested, the alternative is to define new functions. Something like (if we choose to leave it in the class):

```
class Instrument():

    @staticmethod
    def control(check_errors=False):
        pass

class Instr(Instrument):

    @staticmethod
    def checked_control(*args, **kwargs):
        kwargs.setdefault('check_errors', True)
        return Instrument.control(*args, **kwargs)

    prop = checked_control.__func__()
```

It's a very elegant method, but I'm not sure it's much simpler to understand than the repetition for people coming from other instruments. Do you both think I should make the change?

Reply:
What if the new method is called `control` as well and the doc-string of the class tells, that every communication is error checked?
Once you see `Instr.control`, you look up that function and will understand, that you check errors.

Reply:
> In my previous message, I wrote:
> 
> > An alternative would be to add such an option [a `check_error_default` at the class level] to the Instrument class, but I don't know if it would be useful to enough instruments.
> 
> I had not really thought it through, though, and it seams that the use of static methods and class attributes make this impossible.
> 
> As @msmttchr suggested, the alternative is to define new functions. Something like (if we choose to leave it in the class)

@PLLamure , actually I didn't mean to define a dedicated class, but just plain functions at file level.

> 
> It's a very elegant method, but I'm not sure it's much simpler to understand than the repetition for people coming from other instruments. Do you both think I should make the change?

I don't have a strong opinion, my personal taste would be to define the function, but I am ok also to leave the implementation as it is.

Reply:
> What if the new method is called `control` as well and the doc-string of the class tells, that every communication is error checked?
> Once you see `Instr.control`, you look up that function and will understand, that you check errors.

Please don't override existing names and change behaviour like this, that can (iiuc) be quite surprising (and hard to find, needs a docstring lookup). :-)

Just to clairfy, because I don't think that that was the intent: 
If you add methods/other things, please do that in the Instrument class you're currently defining, or, if that is not possible and you need to go one inheritance layer up, in a separate superclass that inherits from `Instrument` (`SMC100Base` or somesuch), but _not_ in `Instrument` itself, because that would change behaviour for all other consumers of that class.

Reply:
@PLLamure, I think, after the discussion there's a general consensus to leave the implementation unchanged (and I also agree with this).